rascaline/src/calculators/neighbor_list.rs

```diff
@@ -0,0 +1,620 @@
+use std::sync::Arc;
+use std::collections::BTreeSet;
+
+use equistore::TensorMap;
+use equistore::{Labels, LabelsBuilder, LabelValue};
+
+use super::CalculatorBase;
+
+use crate::{Error, System};
+
+
+/// This calculator compute the neighbor list for a given spherical cutoff, and
+/// returns the list of distance vectors between pairs of atoms strictly inside
+/// the cutoff.
+///
+/// Users can request either a "full" neighbor list (including an entry for both
+/// `i - j` pairs and `j - i` pairs) or save memory/computational by only
+/// working with "half" neighbor list (only including one entry for each `i/j`
+/// pair)
+///
+/// Self pairs (pairs between an atom and periodic copy itself) can appear when
+/// the cutoff is larger than the cell under periodic boundary conditions. Self
+/// pairs with a distance of 0 are not included in this calculator, even though
+/// they are required when computing SOAP.
+///
+/// This sample produces a single property (`"distance"`) with three components
+/// (`"pair_direction"`) containing the x, y, and z component of the vector from
+/// the first atom in the pair to the second. In addition to the atom indexes,
+/// the samples also contain a pair index, to be able to distinguish between
+/// multiple pairs between the same atom (if the cutoff is larger than the
+/// cell).
+#[derive(Debug, Clone)]
+#[derive(serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+pub struct NeighborList {
+    /// Spherical cutoff to use to determine if two atoms are neighbors
+    pub cutoff: f64,
+    /// Should we compute a full neighbor list (each pair appears twice, once as
+    /// `i-j` and once as `j-i`), or a half neighbor list (each pair only
+    /// appears once)
+    pub full_neighbor_list: bool,
+}
+
+impl CalculatorBase for NeighborList {
+    fn name(&self) -> String {
+        "neighbors list".into()
+    }
+
+    fn parameters(&self) -> String {
+        serde_json::to_string(self).expect("failed to serialize to JSON")
+    }
+
+    fn keys(&self, systems: &mut [Box<dyn System>]) -> Result<Labels, Error> {
+        assert!(self.cutoff > 0.0 && self.cutoff.is_finite());
+
+        if self.full_neighbor_list {
+            FullNeighborList { cutoff: self.cutoff }.keys(systems)
```

Comment:
You probably considered this but are the implementations of `FullNeighborList` and `HalfNeighborList` really so different to not distinguish the cases within one implementation?

From an naive point of view you could just calculate half the neighbor list and stack them with inverted species.

Reply:
Given that three functions need to change depending on the half/full distinction, I feel that having separate structs/impl makes it clearer what's going on.

The issue with doing a half NL and then duplicating, is that we end up doing quite a bit of redundant work (sorting species for the half, only to unsort them later), for a calculator that should be a small wrapper on top of the nl and thus should get out of the way as fast as possible.

Reply:
I see.